### DISCUSSION TITLE
Filter on impossible to match values (null, empty)

### DISCUSSION THREAD

**Discussion by irevoire** [meilisearch/product#539] [Feedback & Feature Proposal]
This discussion comes from this comment on the Meilisearch repository where a user would like to select the fields whose value is equal to `null`: https://github.com/meilisearch/meilisearch/issues/2484#issuecomment-1254995624.

```json
{
  "id": "ignore_me",

  "null": null,
  "array": [],
  "empty": ""
}
```

Currently, all these fields can't be matched by our filters.
1. Is there more « impossible » values currently?
2. What API should we propose to match these?

@mech, since you're the first one asking for this feature, I would love to know how you would like to select these fields.
I was thinking of introducing a new keyword named `IS` that could be followed by `EMPTY` or `NULL` but if there is no other values it feels a little bit overkill :thinking:

**Comment by mech** [comment]
I think if it can mimic SQL like `gender IS NULL` and `gender IS NOT NULL` will be much good enough. My search script actually fallback to Postgres if MeiliSearch instance somehow is offline, so having some parity will be nice. It definitely do not require any value.

If you think about it as a UI, when people select "Is empty", there will be no input field also.

<img width="644" alt="thvvr_21_44_28" src="https://user-images.githubusercontent.com/5850/191763699-60acdc84-82fd-4c63-b454-1cd4e4636ac0.png">

I have another unrelated questions. Since we on the topic of filtering. I do have more filters in mind like:

- Contains
- Does not contain
- Starts with
- Ends with

All copied from Notion filter. I am not sure if you guys plan on that also?

<img width="538" alt="gaf0w_21_46_34" src="https://user-images.githubusercontent.com/5850/191764166-a8832fbe-9b5c-428c-bd0a-c3a25f2513b7.png">

**Reply by gmourier** [reply]
Hey @mech 👋

I've created a fresh [discussion](https://github.com/meilisearch/product/discussions/544) for the `CONTAINS` operator (also mentioning the Starts with, ends with).

**Comment by gmourier** [comment]
> I was thinking of introducing a new keyword named IS that could be followed by EMPTY or NULL but if there is no other values it feels a little bit overkill 🤔

This is purely personal taste, I like the idea of being able to do that as in SQL. These are super common filters for someone having already used a SQL database.

What you are suggesting @irevoire, is to have `NULL` and `EMPTY` instead of `IS NULL` and `IS EMPTY`?

---

@mech Do you need to have both `EMPTY` and `NULL`? 

I'm asking that because maybe having `EMPTY` could match `null` and `''`, `""`. 

Having both brings more versatility but I wonder if it's not introducing some complexity.

**Reply by mech** [reply]
Having `EMPTY` to present just `null` is expected to me. Empty string seems to be presented as having a value traditionally. 

To prevent empty string from being introduced, I would developer will use validation at frontend first, but I could be wrong.

**Reply by FrittenKeeZ** [reply]
I would say `IS NULL` for `null` checks, `IS EMPTY` for `''`, `""`, `[]`, `{}`, and `IS NIL` for either null or empty

**Reply by mech** [reply]
Yup certainty make much sense. Like the way it include empty array and object also 👍

**Comment by FrittenKeeZ** [comment]
The ability to compare two fields could be a nice addition?
Like `field1 > field2`

Various date and time filters would also be very nice, like being able to select users having birthday today, without having to explicit add a field containing only the day and month.

**Reply by irevoire** [reply]
That's interesting; it would be nice if you could open two new product discussions for both features!

**Comment by ahmednfwela** [comment]
I think this is a very important feature, but using the `=` operator would be problematic since the user might expect `field = null` to be equivalent to `field = 'null'` and match `{"field": "null"}`.

I suggest simply introducing new operators `NULL`, `EMPTY` which would have the same syntax rules as `EXISTS`.

---

- `NULL` matches fields that `EXISTS` AND `= NULL`
- `NOT NULL` matches fields that `NOT EXISTS` OR `!= NULL`

1. `{"name":"A", "price": null}`
2. `{"name":"A", "price": 10}`
3. `{"name":"A"}`

`price NULL` would match 1
`price NOT NULL` or `NOT price NULL` would match 2,3
`price EXISTS` would match 1, 2
`price NOT EXISTS` or `NOT price EXISTS` would match 3

common query : `(price EXISTS) AND (price NOT NULL)` would match 2

---

- `EMPTY` matches Array `[ ]` OR Object `{ }` fields that `EXISTS` AND (`= []` OR `= {}`)
- `NOT EMPTY` matches fields that `NOT EXISTS` OR (`!= []` AND `!= {}`)

1. `{"name":"A", "tags": null}`
2. `{"name":"A", "tags": []}`
3. `{"name":"A", "tags": ["hello","world"]}`
4. `{"name":"A"}`
5. `{"name":"A", "tags": {}}`
6. `{"name":"A", "tags": {"t1":"v1"}}`

`tags EMPTY` would match 2,5
`tags NOT EMPTY` or `NOT tags EMPTY` would match 1,3,4,6
`tags NULL` would match 1
`tags NOT NULL` or `NOT tags NULL` would match 2,3,4,5,6
`tags EXISTS` would match 1,2,3,5,6
`tags NOT EXISTS` or `NOT tags EXISTS` would match 4

common query : `(tags EXISTS) AND (tags NOT NULL) AND (tags NOT EMPTY)` would match 3,6

**Reply by irevoire** [reply]
Hey, thanks a lot for your proposition it's super detailed!

There is one thing we won't be able to do, I think (or at least I would like not to do it to stay consistent with the rest).
Currently, `NOT` is an operator; it always works the same way: « it returns the opposite of what the filter matched ».
This means, in this case:
1. {"name":"A", "tags": null}
2. {"name":"A", "tags": []}
3. {"name":"A" }

- `tags NULL` would match `1`
- `tags NOT NULL` or `NOT tags NULL` would match `2` and `3`

Would that be an issue?

-----

Also, not a fan of the `VALID` keyword since it doesn't mean anything; an empty or null value is also « valid ». But I don't have a better proposition 😓 

And finally, shouldn't empty also match `{}`?

**Reply by ahmednfwela** [reply]
actually that would be great, I will modify the original comment to account for this.
the `valid` operator is just a convenience operator that's useful in practice, but personally I wouldn't use it and just use the other explicit operators for readability

**Reply by ahmednfwela** [reply]
however a concern I have is that meili doesn't differentiate between arrays and values in filtering, so an empty array is considered some form of `null`, according to the docs:
![image](https://user-images.githubusercontent.com/63286031/223192231-87408c2e-cd85-47a2-a88f-e10335981489.png)

so what would the `NULL` operator do in the case of empty arrays? 

1. if it matches empty arrays, that would be an unexpected behavior for the users who differentiate between `tags: []` and `tags: null`, because some users might have `tags: null` as a special value meaning all tags, and `tags: []` would mean no tags
2. if it doesn't match empty arrays, it would be the "correct" behavior, but will be inconsistent with meili's strategy of treating arrays and values the same

**Reply by irevoire** [reply]
To me, we should implement 2️⃣, and it’s not really inconsistent since it doesn’t use the equality operator.
The `EXISTS` operator already works with the « field value » instead of the « interpreted content of the field », so I see no issue in following this schema with the new operators we’re talking about 👍

**Reply by ahmednfwela** [reply]
alright then, I updated my original comment, now `EMPTY` only matches empty array fields.
I think the api is more consistent now

**Reply by ahmednfwela** [reply]
> Also, not a fan of the `VALID` keyword since it doesn't mean anything; an empty or null value is also « valid ». But I don't have a better proposition 😓

I agree the word isn't really meaningful and is just a syntax sugar, so I removed it from my comment

> And finally, shouldn't empty also match `{}`?

That's an interesting case I haven't thought about, And yes of course it should match

**Reply by ahmednfwela** [reply]
also I think MongoDB did it best when it comes to nested/array querying, but I think such complex queries would be hard only using text expressions, it would be better if Meili starts adapting filters using json objects which will allow far more complex queries.

e.g.: https://www.mongodb.com/docs/manual/tutorial/query-arrays/

**Reply by Kerollmops** [reply]
The current prototype for the `NULL` keyword is available via Docker.
You can use [the syntax described in the concerned PR](https://github.com/meilisearch/meilisearch/pull/3571#issue-1615449585) here (the same as described in the main message above).

```bash
docker run -it --rm -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:prototype-null-filter-0
```

Please let me know here if anyone has any issue with it 🙏 

⚠️ Prototypes are not recommended for production

**Reply by Kerollmops** [reply]
I published a new prototype version for the `IS NULL` syntax that is available on Docker.
You can use [the syntax described in the concerned PR](https://github.com/meilisearch/meilisearch/pull/3571#issue-1615449585) here (it differs from the one in the main message above, it uses an `IS NULL` and `IS NOT NULL` syntax instead).

```bash
docker run -it --rm -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:prototype-null-filter-1
```

Please let me know here if anyone has any issue with it 🙏 

⚠️ Prototypes are not recommended for production

**Reply by Kerollmops** [reply]
I published a new prototype version for the `IS NULL` and `IS EMPTY` syntax that is available on Docker.
You can use [the syntax described in the concerned PR](https://github.com/meilisearch/meilisearch/pull/3571#issue-1615449585) here.

It differs from the one in the main message above:
 - It uses an `IS NULL` and `IS NOT NULL` syntax.
 - The `IS EMPTY` and `IS NOT EMPTY` matches empty strings `""` too.

```bash
docker run -it --rm -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:prototype-null-empty-filters-0
```

Please let me know here if anyone has any issue with it 🙏

⚠️ Prototypes are not recommended for production

**Reply by mech** [reply]
Just to confirm, this is not in the recent v1.1 releases right.

**Reply by ahmednfwela** [reply]
@mech no I think this will be in v1.2

**Reply by gmourier** [reply]
Yes! @ahmednfwela is right, the release candidate should be available next week

**Comment by macraig** [comment]
Hello everyone 👋 

We've just released the first RC (release candidate) of Meilisearch containing this new feature!
You can test it by using
-  the [release assets](https://github.com/meilisearch/meilisearch/releases)
- the [Meilisearch Docker image](https://hub.docker.com/repository/docker/getmeili/meilisearch/general)

```bash
docker run -it --rm -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:v1.2.0-rc.0
```

You are welcome to leave your feedback in this discussion.

If you encounter any bugs, please report them [here](https://github.com/meilisearch/meilisearch/issues/new?assignees=&labels=&template=bug_report.md&title=).
Thanks in advance for your help and your involvement in Meilisearch ❤️ 

🎉 Official and stable release containing this change will be available on June 5th, 2023

⚠️ RC (release candidates) are not recommended for production

**Comment by macraig** [comment]
Hey folks 👋

[v1.2](https://github.com/meilisearch/meilisearch/releases/tag/v1.2.0) has been released! 🐻‍❄️ We've introduced two new filters:

- `IS EMPTY` matches existing fields with a valid, but empty value

- `IS NULL` matches existing fields with an explicit `null` value 

📚 https://www.meilisearch.com/docs/learn/advanced/filtering#is-null
📚 https://www.meilisearch.com/docs/learn/advanced/filtering#is-empty

**Comment by cklogs** [comment]
@Kerollmops 

I know this is a closed thread, forgive me if this is the wrong spot to ask. I did dig through documentation, github issues and Q&A and could not find exactly this issue being discussed--though it's entirely possible I missed something.

If you'd prefer that I direct the question elsewhere or open a separate discussion, please let me know and I'll promptly update.

For the sake of space efficiency, some documents in my index are intentionally missing fields for which there is no, and will never be a, value.

Initializing all fields for every document (with null or empty) will consume a great deal of extra space, which I would prefer to avoid.

```
[
    {
        "id": "id-01",
        "name": "product 1",
        "specialProductAttribute": [
            "value 1",
            "value 2"
        ]
    },
    {
        "id": "id-02",
        "name": "product 2"
        // specialProductAttribute not defined
    }
]
```

Every single one of these filter expressions will return the product with "id-02".

`"filter": "NOT specialProductAttribute CONTAINS 'value 1'"`

`"filter": "specialProductAttribute != 'value 1'"`

`"filter": "specialProductAttribute IS NOT NULL"`

`"filter": "specialProductAttribute IS NOT EMPTY"`

Is there at all a way to filter out documents that are simple "missing" the targeted field?

Thanks for your help.

**Reply by ahmednfwela** [reply]
I think you are looking for `EXISTS` filter

**Reply by cklogs** [reply]
@ahmednfwela Yes, you are absolutely right. Thank you.

I don't know how that escaped me. Appreciate it.
